ndb/src/google/cloud/ndb/model.py

```diff
@@ -2302,17 +2302,135 @@ def _db_get_value(self, v, unused_p):
 
 
 class DateProperty(DateTimeProperty):
+    """A property that contains :class:`~datetime.date` values.
+
+    .. automethod:: _to_base_type
+    .. automethod:: _from_base_type
+    .. automethod:: _validate
+    """
+
     __slots__ = ()
 
-    def __init__(self, *args, **kwargs):
-        raise NotImplementedError
+    def _validate(self, value):
+        """Validate a ``value`` before setting it.
+
+        Args:
+            value (~datetime.date): The value to check.
+
+        Raises:
+            .BadValueError: If ``value`` is not a :class:`~datetime.date`.
+        """
+        if not isinstance(value, datetime.date):
+            raise exceptions.BadValueError(
+                "Expected date, got {!r}".format(value)
+            )
+
+    def _to_base_type(self, value):
+        """Convert a value to the "base" value type for this property.
+
+        Args:
+            value (~datetime.date): The value to be converted.
+
+        Returns:
+            ~datetime.datetime: The converted value: a datetime object with the
+            time set to ``00:00``.
+
+        Raises:
+            TypeError: If ``value`` is not a :class:`~datetime.date`.
+        """
+        if not isinstance(value, datetime.date):
+            raise TypeError(
+                "Cannot convert to datetime expected date value; "
+                "received {}".format(value)
+            )
+        return datetime.datetime(value.year, value.month, value.day)
+
+    def _from_base_type(self, value):
+        """Convert a value from the "base" value type for this property.
+
+        Args:
+            value (~datetime.datetime): The value to be converted.
+
+        Returns:
+            ~datetime.date: The converted value: the date that ``value``
+            occurs on.
+        """
+        return value.date()
```

Comment:
There [was][1] an `assert` here but I removed it. ISTM that the base type is coming from the storage layer (i.e. not from users) so the caller won't mess this up.

[1]: https://github.com/GoogleCloudPlatform/datastore-ndb-python/blob/cf4cab3f1f69cd04e1a9229871be466b53729f3f/ndb/model.py#L2187

Reply:
Yep. Same as with discussion on `JSONProperty`, assuming the passed value is `bytes` there.

Reply:
Did you mean `bytes` or is that a "brain typo" (since `bytes` is the type relevant to `JSONProperty`)?

Reply:
It wasn't a typo, but maybe unclear language. I'm just saying that, yes, the same reasoning that allows us to assume we're getting a `datetime.datetime` here is the same that allows us to assume we're geting `bytes` in JSONProperty.

Reply:
:+1: my bad